File: src/ProtectManager/component/PM_System_Base_Common_Service/src/pm-common/src/main/java/openbackup/system/base/common/model/job/Job.java
/*
* This file is a part of the open-eBackup project.
* This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
* If a copy of the MPL was not distributed with this file, You can obtain one at
* http://mozilla.org/MPL/2.0/.
*
* Copyright (c) [2024] Huawei Technologies Co.,Ltd.
*
* THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
* EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
* MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
*/
package openbackup.system.base.common.model.job;

import openbackup.system.base.common.utils.JSONObject;
import openbackup.system.base.query.PageQueryConfig;
import openbackup.system.base.sdk.job.model.request.JobMessage;

import com.baomidou.mybatisplus.annotation.TableId;
import com.baomidou.mybatisplus.annotation.TableName;

import lombok.Data;

import java.util.Optional;

/**
 * job DAO类
 *
 */
@TableName(value = "t_job")
@Data
@PageQueryConfig(
        conditions = {"%source_name%", "%source_location%", "%target_name%", "%target_location%"},
        orders = {"start_time", "copy_time", "end_time"})
public class Job {
    @TableId
    private String jobId;

    private String userId;

    private String type;

    private Integer progress;

    private Long startTime;

    private Long endTime;

    private Long lastUpdateTime;

    private String status;

    private String parentId;

    private String speed;

    private String detail;

    private String detailPara;

    private String extendStr;

    private String associativeId;

    /**
     * 保护对象ID
     */
    private String sourceId;

    /**
     * 保护对象名称
     */
    private String sourceName;

    /**
     * 资源类型
     */
    private String sourceType;

    /**
     * 资源子类型
     */
    private String sourceSubType;

    /**
     * 保护对象位置
     */
    private String sourceLocation;

    /**
     * 目标名称
     */
    private String targetName;

    /**
     * 目标对象位置
     */
    private String targetLocation;

    /**
     * 副本ID
     */
    private String copyId;

    /**
     * 副本创建时间
     */
    private Long copyTime;

    /**
     * 任务能否被中止
     */
    private Boolean enableStop;

    /**
     * 是否为系统任务
     */
    private Boolean isSystem;

    /**
     * 请求ID
     */
    private String requestId;

    /**
     * 是否可见
     */
    private Boolean isVisible;

    /**
     * 附加状态
     */
    private String additionalStatus;

    private String message;

    private String data;

    /**
     * 任务详情是否包含特定级别事件标志位。
     * 使用整数的二进制形式表示多个标志位。
     * 目前表示3个标志位，从高位到低位分别表示 “是否包含警告事件”、“是否包含错误事件”、“是否包含严重事件”。
     * 0表示不包含，1表示包含包含，示例：（0b）110 表示包含警告、错误事件，不包含严重事件，jobLogLevel此时为6。
     */
    private Integer logLevels;

    /**
     * 任务所在节点esn
     */
    private String deviceEsn;

    /**
     * 存储单元id
     */
    private String storageUnitId;

    /**
     * 备份任务slaId
     */
    private String slaId;

    /**
     * 演练计划任务及其子任务存储演练id
     */
    private String exerciseId;

    /**
     * 演练计划的子任务存储父任务id
     */
    private String exerciseJobId;

    /**
     * 资源组id
     */
    private String resourceGroupId;

    /**
     * 资源组备份父任务id
     */
    private String groupBackupJobId;

    /**
     * mark，用户标记的处理意见
     */
    private String mark;

    /**
     * markStatus，任务的处理状态
     */
    private String markStatus;

    /**
     * parse message as JobMessage
     *
     * @return JobMessage
     */
    public JobMessage message() {
        return JSONObject.fromObject(message).toBean(JobMessage.class);
    }

    /**
     * get job message payload with traffic data
     *
     * @return job message payload
     */
    public JSONObject resolvePayloadAndTraffic() {
        JSONObject json = JSONObject.fromObject(message);
        JSONObject payload = Optional.ofNullable(json.getJSONObject("payload")).orElseGet(JSONObject::new);
        JSONObject traffic = json.getJSONObject("traffic");
        if (traffic != null) {
            payload.putAll(traffic);
        }
        if (associativeId != null) {
            payload.put("request_id", associativeId);
        }
        return payload;
    }
}

File: src/ProtectManager/component/PM_Data_Protection_Service/src/app/ai_sorter/utils/job.py
# This file is a part of the open-eBackup project.
# This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
# If a copy of the MPL was not distributed with this file, You can obtain one at
# http://mozilla.org/MPL/2.0/.
#
# Copyright (c) [2024] Huawei Technologies Co.,Ltd.
#
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
# EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
# MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
import datetime
import json
from typing import List

from app.ai_sorter.model import VM
from app.common.schemas.job_schemas import JobSchema
from app.protection.object.models.projected_object import ProtectedObject
from app.common.toolkit import query_job_list
from app.common import logger

log = logger.get_logger(__name__)


def get_backup_jobs_per_vm(vm_obj: ProtectedObject, min_jobs_per_vm: int) -> \
        List[JobSchema]:
    job_list = []
    job_condition_map = {
        'sourceName': vm_obj.name,
        'statusList': ['SUCCESS'],
        'types': ['BACKUP'],
        'orderBy': 'startTime',
        'orderType': 'asc',
        'pageSize': 200
    }
    response = query_job_list(job_condition_map)
    if response is None:
        return job_list

    paged_job_bos = json.loads(response)
    total_count = paged_job_bos['totalCount']
    if total_count < min_jobs_per_vm:
        return job_list

    for job in paged_job_bos['records']:
        job_schema = JobSchema(**job)
        job_list.append(job_schema)
    return job_list


def get_job_speed(job: JobSchema):
    """
    返回指定任务的吞吐量

    :param job: 任务
    :return: 吞吐量，MB/s; -1则为失败。
    """
    job_size = parse_job_size(job)
    if job_size == -1:
        return -1
    job_time_range = (job.end_time - job.start_time) / 1000
    try:
        return job_size / max(job_time_range, 0.1)
    except ZeroDivisionError:
        log.info(f"job_time_range is near zero, "
                 f"end_time: {job.end_time}, start_time:{job.start_time}")
        return job_size / 0.1


def parse_job_size(job_inf: JobSchema):
    """
    解析job size数据，并全部统一为MB为单位的大小
    :param job_inf: job信息数据
    :return: 通过远程调用查询成功备份的历史任务
    """
    prev_job_size_extend_str = json.loads(job_inf.extend_str)
    try:
        prev_job_size_inf = prev_job_size_extend_str['dataAfterReduction']
    except KeyError:
        log.info(f"read job size error! no such key:[dataAfterReduction] in "
                 f"extend string!")
        return -1

    # 将job size统一转化为MB
    if 'TB' in prev_job_size_inf:
        prev_job_size = float(prev_job_size_inf.rstrip('TB').rstrip()) * 1024 * 1024
    elif 'GB' in prev_job_size_inf:
        prev_job_size = float(prev_job_size_inf.rstrip('GB').rstrip()) * 1024
    elif 'MB' in prev_job_size_inf:
        prev_job_size = float(prev_job_size_inf.rstrip('MB').rstrip())
    elif 'KB' in prev_job_size_inf:
        prev_job_size = float(prev_job_size_inf.rstrip('KB').rstrip()) / 1024
    elif 'B' in prev_job_size_inf:
        prev_job_size = float(prev_job_size_inf.rstrip('B').rstrip()) / 1024 / 1024
    else:
        log.info(
            f'error, unsupported unit in [{prev_job_size_inf}]! for job size')
        prev_job_size = -1

    return prev_job_size


def get_prev_jobs(vm_item: VM, window_len: int) -> List[JobSchema]:
    """
    获取成功备份的历史任务
    :param window_len: 备份历史成功任务数目
    :param vm_item: VM类，含job调度信息
    :return: 通过远程调用查询成功备份的历史任务
    """
    response = query_job_list(
        {'fromStartTime': int(
            datetime.datetime.utcnow().timestamp()
            - float(vm_item.schedule_freq * window_len)) * 1000,
         'sourceName': vm_item.name,
         'statusList': ['SUCCESS'], 'types': ['BACKUP'],
         'orderBy': 'startTime', 'orderType': 'asc'})

    paged_job_bos = json.loads(response)
    job_list = []
    for job in paged_job_bos['records']:
        job_schema = JobSchema(**job)
        job_list.append(job_schema)

    return job_list


def get_prev_job_speed_list(vm_item: VM, window_len: int) -> List[float]:
    pre_job_list = get_prev_jobs(vm_item, window_len)

    prev_job_speed_list = []
    for prev_job in pre_job_list:
        job_speed = get_job_speed(prev_job)
        if job_speed < 0:
            continue
        prev_job_speed_list.append(job_speed)
    return prev_job_speed_list
